src/TorchSharp/Data/DataLoader.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Linq;
+using System.Collections;
+using System.Collections.Generic;
+
+using static TorchSharp.torch;
+
+namespace TorchSharp.Data
+{
+    public class DataLoader: IEnumerable<(Tensor, Tensor)>
+    {
+        private Dataset dataset;
+        private int batchSize;
+        private bool shuffle;
+        private Device device;
+
+        public DataLoader(Dataset dataset, int batchSize, bool shuffle = false, Device device = null)
+        {
+            this.dataset = dataset;
+            this.batchSize = batchSize;
+            this.shuffle = shuffle;
+            this.device = device ?? CPU;
+        }
+        public IEnumerator<(Tensor, Tensor)> GetEnumerator()
+        {
+            return new DataLoaderEnumerator(dataset, batchSize, shuffle, device);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public long Count => (dataset.Count() - 1) / batchSize + 1;
+
+        private class DataLoaderEnumerator : IEnumerator<(Tensor, Tensor)>
+        {
+            private Dataset dataset;
+            private IEnumerator<object> load;
+            private int batchSize;
+            private Device device;
+            public DataLoaderEnumerator(Dataset dataset, int batchSize, bool shuffle, Device device)
+            {
+                this.dataset = dataset;
+                this.batchSize = batchSize;
+                this.device = device;
+
+                if (shuffle)
+                {
+                    var r = new Random();
+                    load = dataset.GetDataEnumerable()
+                        .Select(x => new {Number = r.Next(), Item = x})
+                        .OrderBy(x => x.Number)
+                        .Select(x => x.Item)
+                        .GetEnumerator();
+                }
+                else load = dataset.GetDataEnumerable().GetEnumerator();
+            }
+
+            private Tensor dataTensor;
+            private Tensor labelTensor;
+            private (Tensor, Tensor) tmp;
```

Comment:
remove these members, there are not necessary

Reply:
dataTensor, labelTensor is required for return current value.

Reply:
I change that to directly change Current value.